src/sentry/reprocessing2.py

```diff
@@ -347,73 +351,94 @@ def _get_info_reprocessed_key(group_id):
+    # contain {}.
+    key = f"re2:remaining:{{{project_id}:{old_group_id}}}"
+
+    llen = None
 
     if datetime_to_event:
-        client.lpush(
+        llen = client.lpush(
             key,
             *(f"{to_timestamp(datetime)};{event_id}" for datetime, event_id in datetime_to_event),
         )
         client.expire(key, settings.SENTRY_REPROCESSING_SYNC_TTL)
 
-    if (
-        force_flush_batch
-        or client.llen(key) > settings.SENTRY_REPROCESSING_REMAINING_EVENTS_BUF_SIZE
-    ):
-        event_ids_batch = []
-        min_datetime = None
-        max_datetime = None
-
-        # TODO: Redis 6 introduces LPOP with <count> argument, use here?
-        while True:
-            row = client.lpop(key)
-            if row is None:
-                break
+    MAX_LLEN = settings.SENTRY_REPROCESSING_REMAINING_EVENTS_BUF_SIZE
 
-            datetime_raw, event_id = row.split(";")
-            datetime = to_datetime(float(datetime_raw))
+    if force_flush_batch or (llen is not None and llen > MAX_LLEN) or client.llen(key) > MAX_LLEN:
+        new_key = f"{key}:{uuid.uuid4().hex}"
 
-            assert datetime is not None
-
-            if min_datetime is None or datetime < min_datetime:
-                min_datetime = datetime
-            if max_datetime is None or datetime > max_datetime:
-                max_datetime = datetime
-
-            event_ids_batch.append(event_id)
+        try:
+            # Rename `key` to a new temp key that is passed to celery task. We
+            # use `renamenx` instead of `rename` only to detect UUID collisions.
+            assert client.renamenx(key, new_key), "UUID collision for new_key?"
+        except redis.exceptions.ResponseError:
+            # `key` does not exist in Redis. `ResponseError` is a bit too broad
+            # but it seems we'd have to do string matching on error message
+            # otherwise.
+            return
 
         from sentry.tasks.reprocessing2 import handle_remaining_events
 
         handle_remaining_events.delay(
             project_id=project_id,
             old_group_id=old_group_id,
             new_group_id=new_group_id,
-            event_ids=event_ids_batch,
             remaining_events=remaining_events,
-            from_timestamp=min_datetime,
-            to_timestamp=max_datetime,
+            event_ids_redis_key=new_key,
         )
 
 
+def pop_remaining_event_ids_from_redis(key):
+    client = _get_sync_redis_client()
+    event_ids_batch = []
+    min_datetime = None
+    max_datetime = None
+
+    for row in client.lrange(key, 0, -1):
+        datetime_raw, event_id = row.split(";")
+        datetime = to_datetime(float(datetime_raw))
+
+        assert datetime is not None
+
+        if min_datetime is None or datetime < min_datetime:
+            min_datetime = datetime
+        if max_datetime is None or datetime > max_datetime:
+            max_datetime = datetime
+
+        event_ids_batch.append(event_id)
+
+    client.delete(key)
```

Comment:
`client.lpop()` was atomic, right? Is there any danger of two task instances running at the same time and working on the same `lrange`?

Reply:
Since I now rename the key to something unique I don't have concurrent access on it anymore.